File: src/main/java/com/example/game/units/Shuttle.java
package com.example.game.units;

import com.example.game.core.Point2D;
import com.example.game.core.Unit;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Shuttle extends Unit implements Transporter {
    private final int capacity = 8;
    private final List<Unit> passengers = new ArrayList<>();

    public Shuttle(int id, Point2D startPos) {
        super("셔틀", id, true, startPos);
    }

    @Override
    public int getCapacity() { return capacity; }

    @Override
    public int getPassengerCount() { return passengers.size(); }

    @Override
    public boolean board(Unit u) {
        if (u == null) return false;
        if (passengers.size() >= capacity) {
            log("탑승 실패: 정원 초과입니다. (" + u.getDisplayName() + ")");
            return false;
        }
        // Knight/Archer만 탑승 가능
        if (!(u instanceof Knight) && !(u instanceof Archer)) {
            log("탑승 실패: " + u.getDisplayName() + " 는 탑승할 수 없습니다.");
            return false;
        }
        passengers.add(u);
        log(u.getDisplayName() + " 탑승 완료. (현재 " + passengers.size() + "/" + capacity + ")");
        // 탑승 즉시 좌표 동기화
        setUnitPositionSilently(u, this.position);
        return true;
    }

    @Override
    public List<Unit> unloadAll() {
        if (passengers.isEmpty()) {
            log("하차할 승객이 없습니다.");
            return Collections.emptyList();
        }
        List<Unit> out = new ArrayList<>(passengers);
        for (Unit u : out) {
            setUnitPositionSilently(u, this.position);
            log(u.getDisplayName() + " 하차합니다. (좌표 " + this.position + ")");
        }
        passengers.clear();
        return out;
    }

    @Override
    public void moveTo(Point2D dest) {
        log("날아서 " + position + " -> " + dest + " 로 이동합니다. (승객 " + passengers.size() + "명)");
        this.position = dest;
        // 승객 좌표도 동기화
        for (Unit u : passengers) {
            setUnitPositionSilently(u, dest);
        }
    }

    @Override
    protected boolean canAttack(Unit target) {
        // 셔틀은 공격 불가
        return false;
    }

    private static void setUnitPositionSilently(Unit u, Point2D p) {
        try {
            Field f = Unit.class.getDeclaredField("position");
            f.setAccessible(true);
            f.set(u, p);
        } catch (Exception ignored) { }
    }
}

File: src/main/kotlin/com/example/game/units/Shuttle.kt
package com.example.game.units

import com.example.game.core.Point2D
import com.example.game.core.UnitBase
import com.example.game.traits.Movable
import com.example.game.traits.Transporter

/**
 * Shuttle(셔틀): Knight/Archer만 탑승 가능한 비행 수송 유닛. 정원은 8.
 */
class Shuttle(id: Int, start: Point2D) :
    UnitBase(id = id, displayName = "셔틀", position = start, isFlying = true),
    Movable, Transporter {

    override val capacity: Int = 8
    private val passengers = mutableListOf<UnitBase>()

    override fun moveTo(dest: Point2D) {
        log("하늘을 날아 $position → $dest 로 이동합니다. (탑승자 ${passengers.size}명 동반)")
        position = dest
    }

    override fun board(unit: UnitBase): Boolean {
        val allowed = unit is Knight || unit is Archer
        if (!allowed) {
            log("탑승 실패: ${unit.displayName}#${unit.id} 은(는) 탑승할 수 없습니다.")
            return false
        }
        if (passengers.size >= capacity) {
            log("탑승 실패: 정원 초과(${capacity}).")
            return false
        }
        passengers += unit
        log("${unit.displayName}#${unit.id} 탑승 완료. (현재 ${passengers.size}/${capacity})")
        return true
    }

    override fun disembarkAll(): List<UnitBase> {
        if (passengers.isEmpty()) {
            log("하차할 승객이 없습니다.")
            return emptyList()
        }
        val out = passengers.toList()
        passengers.clear()
        log("승객 ${out.size}명 전원 하차.")
        return out
    }
}